### PR TITLE
Start porting the lowering passes to the new pass manager

### PR DESCRIPTION
**Pull request by bsaleil**
This patch starts porting the lowering passes added in SpirvLower::addPasses to the
new pass manager. It also makes some changes to the infrastructure for the new
pass manager to allow adding function analyses.

### CODE REVIEW

**Comment by amdvlk-admin** [comment]
### Test summary for commit a7b8f14178636a902aca1c89e4f5cda52bc88bac
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 93d6aba82af14ae57ae6d84a4cf0cca3d9aa8c6b</li>
<li>SPVGEN: dev 79f8e32dd7242d7d34fa0074f761c9d794f11811</li>
<li>XGL: dev 5c3c2dd430222cdf1b46c9d385e9c6c5e89604e1</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 0cb96bc82ca13e0fef3684a4ac8f6eff50bce257</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/228128)</summary><ul><li>Built with version 1.2.5.2</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 31146/57032 (54.6%)</li>
<li>Failed: 0/57032 (0.0%)</li>
<li>Not Supported: 25886/57032 (45.4%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 31145/57032 (54.6%)</li>
<li>Failed: 0/57032 (0.0%)</li>
<li>Not Supported: 25887/57032 (45.4%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 31459/57032 (55.2%)</li>
<li>Failed: 0/57032 (0.0%)</li>
<li>Not Supported: 25573/57032 (44.8%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx103</summary><ul>

<li>Passed: 34853/57032 (61.1%)</li>
<li>Failed: 0/57032 (0.0%)</li>
<li>Not Supported: 22179/57032 (38.9%)</li>
</ul></details>


</ul></details>

**Comment by JaxLinAMD** [comment]
@piotrAMD PING please help review this PR.

File: lgc/interface/lgc/PassManager.h [hunk 1]
```diff
@@ -51,8 +51,14 @@ class PassManager : public llvm::ModulePassManager {
 public:
   static PassManager *Create();
   virtual ~PassManager() {}
+  template <typename PassBuilderT> bool registerFunctionAnalysis(PassBuilderT &&PassBuilder) {
+    return functionAnalysisManager.registerPass(PassBuilder);
```
Comment (kuhar) [inline lgc/interface/lgc/PassManager.h:55]: Why is PassBuilder taken by `&&` but not forwarded? I think this should either take it by `&` or use `std::forward`

**Comment by amdvlk-admin** [comment]
### Test summary for commit 89177fcf3391dc90ac2b031908d5132d38741499
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 93d6aba82af14ae57ae6d84a4cf0cca3d9aa8c6b</li>
<li>SPVGEN: dev 79f8e32dd7242d7d34fa0074f761c9d794f11811</li>
<li>XGL: dev 0df9173340136b73ff2aa49647d18b8acd43ca5f</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 0cb96bc82ca13e0fef3684a4ac8f6eff50bce257</li>
</ul></details>

<details><summary>CTS tests (Failed: 1/188379)</summary><ul><li>Built with version 1.2.5.2</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 31242/57279 (54.5%)</li>
<li>Failed: 0/57279 (0.0%)</li>
<li>Not Supported: 26037/57279 (45.5%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 10184/16542 (61.6%)</li>
<li><details><summary>Failed: 1/16542 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.binding_model.buffer_device_address.set3.depth3.basessbo.convertchecku64.store.replay.std140.frag
Stack trace: Script:
Crash

```
</p></details></li>

<li>Not Supported: 6357/16542 (38.4%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 31555/57279 (55.1%)</li>
<li>Failed: 0/57279 (0.0%)</li>
<li>Not Supported: 25724/57279 (44.9%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx103</summary><ul>

<li>Passed: 34949/57279 (61.0%)</li>
<li>Failed: 0/57279 (0.0%)</li>
<li>Not Supported: 22330/57279 (39.0%)</li>
</ul></details>


</ul></details>

File: llpc/lower/llpcSpirvLower.cpp [hunk 2]
```diff
@@ -138,6 +140,34 @@ void SpirvLower::removeConstantExpr(Context *context, GlobalVariable *global) {
...
+  // we marked (non-entrypoint) functions as "always inline" just after SPIR-V reading.
+  passMgr.addPass(AlwaysInlinerPass());
+  passMgr.addPass(GlobalDCEPass());
+
+  // Lower SPIR-V access chain
+  passMgr.addPass(SpirvLowerAccessChain());
+
+  // TODO Only some passes have been ported to the new pass manager. Trying to add
+  // the lowering passes to a new pass manager results in a fatal error for now.
+  report_fatal_error("The new pass manager is not fully implemented yet.");
```
Comment (piotrAMD) [inline llpc/lower/llpcSpirvLower.cpp:168]: Would it make sense to move the fatal error to llpcCompiler.cpp after runPasses()? This way we could actually run the passes ported already.

Comment (bsaleil) [reply llpc/lower/llpcSpirvLower.cpp:168]: Sure, I can do that. I already tried running the passes though. The passes successfully ran and I observed only minor changes in the code, but I'll wait until I port more passes to do more extensive testing.

**Comment by kuhar** [comment]
@bsaleil are you added as a collaborator to this repo?

/rebase

**Review by piotrAMD — APPROVED** [review]

**Comment by amdvlk-admin** [comment]
### Test summary for commit 22f2a94f5a6e3847ccace2ae9826ba95041dcdeb
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 652389aacbec397eeac3635ecbf6dc538251780c</li>
<li>SPVGEN: dev 79f8e32dd7242d7d34fa0074f761c9d794f11811</li>
<li>XGL: dev 0df9173340136b73ff2aa49647d18b8acd43ca5f</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 0a3fa7fcc965c8e6d37ad16469bbf4ea8caf90ee</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/229120)</summary><ul><li>Built with version 1.2.5.2</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 31242/57280 (54.5%)</li>
<li>Failed: 0/57280 (0.0%)</li>
<li>Not Supported: 26038/57280 (45.5%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 31241/57280 (54.5%)</li>
<li>Failed: 0/57280 (0.0%)</li>
<li>Not Supported: 26039/57280 (45.5%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 31555/57280 (55.1%)</li>
<li>Failed: 0/57280 (0.0%)</li>
<li>Not Supported: 25725/57280 (44.9%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx103</summary><ul>

<li>Passed: 34949/57280 (61.0%)</li>
<li>Failed: 0/57280 (0.0%)</li>
<li>Not Supported: 22331/57280 (39.0%)</li>
</ul></details>


</ul></details>

**Comment by bsaleil** [comment]
> @bsaleil are you added as a collaborator to this repo?
> 
> /rebase

I'm actually not sure but I think so.

**Comment by kuhar** [comment]
I think you need to ask somebody to add you (@jinjianrong ?):

image
image

**Comment by amdvlk-admin** [comment]
### Test summary for commit 8b831a72e98a8f2810bca0b4cb7b49615e1dc63d
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev 652389aacbec397eeac3635ecbf6dc538251780c</li>
<li>SPVGEN: dev 79f8e32dd7242d7d34fa0074f761c9d794f11811</li>
<li>XGL: dev 0df9173340136b73ff2aa49647d18b8acd43ca5f</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev 0a3fa7fcc965c8e6d37ad16469bbf4ea8caf90ee</li>
</ul></details>

<details><summary>CTS tests (Failed: 0/229120)</summary><ul><li>Built with version 1.2.5.2</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 31242/57280 (54.5%)</li>
<li>Failed: 0/57280 (0.0%)</li>
<li>Not Supported: 26038/57280 (45.5%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 31241/57280 (54.5%)</li>
<li>Failed: 0/57280 (0.0%)</li>
<li>Not Supported: 26039/57280 (45.5%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 31555/57280 (55.1%)</li>
<li>Failed: 0/57280 (0.0%)</li>
<li>Not Supported: 25725/57280 (44.9%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx103</summary><ul>

<li>Passed: 34949/57280 (61.0%)</li>
<li>Failed: 0/57280 (0.0%)</li>
<li>Not Supported: 22331/57280 (39.0%)</li>
</ul></details>


</ul></details>

**Pull request merged**